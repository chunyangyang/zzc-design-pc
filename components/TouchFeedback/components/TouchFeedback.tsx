import React, { PureComponent } from 'react';
import classNames from 'classnames';
import { isFunction } from '../../_util/typeof';
import { TouchFeedbackProps } from '../propstype';

export default class TouchFeedback extends PureComponent<TouchFeedbackProps, any> {
    constructor( props ) {
        super( props );
    }
    state = {
        active: false
    }
    static defaultProps = {
        activeStyle: null,
        activeClassName: '',
        disabled: false
    }

    triggerEvent ( type: string, isActive: boolean, e: Event ): void {
        // 获取当前触发事件
        // 当有传入对应事件的函数则会触发
        const eventType = `on${type}`;
        if ( this.props[eventType] ) {
            isFunction( this.props[eventType] ) && this.props[eventType]( e );
        }
        if ( isActive !== this.state.active ) {
            this.setState( {
                active: isActive
            } );
        }
    }

    onTouchStart = ( e: Event ): void => {
        this.triggerEvent( 'TouchStart', true, e );
    }

    onTouchMove = ( e: Event ): void => {
        this.triggerEvent( 'TouchMove', false, e );
    }

    onTouchEnd = ( e: Event ): void => {
        this.triggerEvent( 'TouchEnd', false, e );
    }

    onTouchCancel = ( e: Event ): void => {
        this.triggerEvent( 'TouchCancel', false, e );
    }

    onMouseDown = ( e: Event ): void => {
        // 当在pc上传入TouchStart事件会触发事件函数
        if ( this.props.onTouchStart ) {
            this.triggerEvent( 'TouchStart', true, e );
        }
        this.triggerEvent( 'MouseDown', true, e );
    }

    onMouseUp = ( e: Event ): void => {
        // 当在pc上传入TouchEnd事件会触发事件函数
        if ( this.props.onTouchEnd ) {
            this.triggerEvent( 'TouchEnd', true, e );
        }
        this.triggerEvent( 'MouseUp', false, e );
    }

    // m端按下后手指划出元素触发取消touch
    onMouseLeave = ( e: Event ): void => {
        this.triggerEvent( 'MouseLeave', false, e );
    }

    render() {
        const { children, activeStyle, activeClassName, disabled } = this.props;
        const events = {
            onTouchStart: this.onTouchStart,
            onTouchMove: this.onTouchMove,
            onTouchEnd: this.onTouchEnd,
            onTouchCancel: this.onTouchCancel,
            onMouseDown: this.onMouseDown,
            onMouseUp: this.onMouseUp,
            onMouseLeave: this.onMouseLeave
        };

        const child = React.Children.only( children );

        if ( !disabled && this.state.active ) {
            let { style, className } = child.props;

            // 合拼style
            if ( activeStyle ) {
                style = { ...style, ...activeStyle };
            }

            // 生成新classname
            className = classNames( className, activeClassName );
            // 返回一个带有touch的样式和classname的元素
            return React.cloneElement( child, {
                className,
                style,
                ...events
            } );
        }

        // 返回绑定事件的元素
        return React.cloneElement( child, events );
    }
}