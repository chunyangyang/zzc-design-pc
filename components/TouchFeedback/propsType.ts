export interface TouchFeedbackProps {
    activeStyle?: React.CSSProperties
    activeClassName?: string,
    disabled?: boolean,
    onTouchStart?: Function;
    onTouchEnd?: Function;
    onTouchCancel?: Function;
    onMouseDown?: Function;
    onMouseUp?: Function;
    onMouseLeave?: Function;
}