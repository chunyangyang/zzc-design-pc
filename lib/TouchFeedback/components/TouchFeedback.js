'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _typeof = require('../../_util/typeof');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TouchFeedback = function (_PureComponent) {
    (0, _inherits3.default)(TouchFeedback, _PureComponent);

    function TouchFeedback(props) {
        (0, _classCallCheck3.default)(this, TouchFeedback);

        var _this = (0, _possibleConstructorReturn3.default)(this, (TouchFeedback.__proto__ || (0, _getPrototypeOf2.default)(TouchFeedback)).call(this, props));

        _this.state = {
            active: false
        };
        _this.onTouchStart = function (e) {
            _this.triggerEvent('TouchStart', true, e);
        };
        _this.onTouchMove = function (e) {
            _this.triggerEvent('TouchMove', false, e);
        };
        _this.onTouchEnd = function (e) {
            _this.triggerEvent('TouchEnd', false, e);
        };
        _this.onTouchCancel = function (e) {
            _this.triggerEvent('TouchCancel', false, e);
        };
        _this.onMouseDown = function (e) {
            // 当在pc上传入TouchStart事件会触发事件函数
            if (_this.props.onTouchStart) {
                _this.triggerEvent('TouchStart', true, e);
            }
            _this.triggerEvent('MouseDown', true, e);
        };
        _this.onMouseUp = function (e) {
            // 当在pc上传入TouchEnd事件会触发事件函数
            if (_this.props.onTouchEnd) {
                _this.triggerEvent('TouchEnd', true, e);
            }
            _this.triggerEvent('MouseUp', false, e);
        };
        // m端按下后手指划出元素触发取消touch
        _this.onMouseLeave = function (e) {
            _this.triggerEvent('MouseLeave', false, e);
        };
        return _this;
    }

    (0, _createClass3.default)(TouchFeedback, [{
        key: 'triggerEvent',
        value: function triggerEvent(type, isActive, e) {
            // 获取当前触发事件
            // 当有传入对应事件的函数则会触发
            var eventType = 'on' + type;
            if (this.props[eventType]) {
                (0, _typeof.isFunction)(this.props[eventType]) && this.props[eventType](e);
            }
            if (isActive !== this.state.active) {
                this.setState({
                    active: isActive
                });
            }
        }
    }, {
        key: 'render',
        value: function render() {
            var _props = this.props,
                children = _props.children,
                activeStyle = _props.activeStyle,
                activeClassName = _props.activeClassName,
                disabled = _props.disabled;

            var events = {
                onTouchStart: this.onTouchStart,
                onTouchMove: this.onTouchMove,
                onTouchEnd: this.onTouchEnd,
                onTouchCancel: this.onTouchCancel,
                onMouseDown: this.onMouseDown,
                onMouseUp: this.onMouseUp,
                onMouseLeave: this.onMouseLeave
            };
            var child = _react2.default.Children.only(children);
            if (!disabled && this.state.active) {
                var _child$props = child.props,
                    style = _child$props.style,
                    className = _child$props.className;
                // 合拼style

                if (activeStyle) {
                    style = (0, _assign2.default)({}, style, activeStyle);
                }
                // 生成新classname
                className = (0, _classnames2.default)(className, activeClassName);
                // 返回一个带有touch的样式和classname的元素
                return _react2.default.cloneElement(child, (0, _assign2.default)({ className: className,
                    style: style }, events));
            }
            // 返回绑定事件的元素
            return _react2.default.cloneElement(child, events);
        }
    }]);
    return TouchFeedback;
}(_react.PureComponent);

exports.default = TouchFeedback;

TouchFeedback.defaultProps = {
    activeStyle: null,
    activeClassName: '',
    disabled: false
};