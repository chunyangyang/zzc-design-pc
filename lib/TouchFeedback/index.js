'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TouchFeedback = require('./components/TouchFeedback');

var _TouchFeedback2 = _interopRequireDefault(_TouchFeedback);

var _Error = require('../_util/Error');

var _Error2 = _interopRequireDefault(_Error);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TouchFeedbackEntrance = function (_PureComponent) {
    (0, _inherits3.default)(TouchFeedbackEntrance, _PureComponent);

    function TouchFeedbackEntrance() {
        (0, _classCallCheck3.default)(this, TouchFeedbackEntrance);
        return (0, _possibleConstructorReturn3.default)(this, (TouchFeedbackEntrance.__proto__ || (0, _getPrototypeOf2.default)(TouchFeedbackEntrance)).apply(this, arguments));
    }

    (0, _createClass3.default)(TouchFeedbackEntrance, [{
        key: 'render',
        value: function render() {
            return _react2.default.createElement(
                _Error2.default,
                { componentName: 'TouchFeedback' },
                _react2.default.createElement(_TouchFeedback2.default, this.props)
            );
        }
    }]);
    return TouchFeedbackEntrance;
}(_react.PureComponent);

exports.default = TouchFeedbackEntrance;