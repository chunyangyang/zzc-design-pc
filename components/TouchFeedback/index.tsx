import React, { PureComponent } from 'react';
import TouchFeedback from './components/TouchFeedback';
import Error from '../_util/Error';
import { TouchFeedbackProps } from './propstype';

export default class TouchFeedbackEntrance extends PureComponent<TouchFeedbackProps, any> {
    render() {
        return (
            <Error componentName='TouchFeedback'>
                <TouchFeedback {...this.props} />
            </Error>
        );
    }
}