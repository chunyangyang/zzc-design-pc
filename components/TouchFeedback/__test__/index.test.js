import React from 'react';
import { mount, shallow, render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import TouchFeedback from '../index.tsx';

test( 'Button', () => {
    const wrapper = render( <div>
        <TouchFeedback
            activeClassName='active-class'
        >
            <div className='touch-btn'>点击添加ClassName</div>
        </TouchFeedback>
        <TouchFeedback
            activeStyle={{ background: 'red' }}
        >
            <div className='touch-btn'>点击添加Style</div>
        </TouchFeedback>
        <TouchFeedback
            activeClassName='active-class'
            disabled
        >
            <div className='touch-btn'>禁止触发touch</div>
        </TouchFeedback>
    </div> );
    expect( renderToJson( wrapper ) ).toMatchSnapshot();
} );