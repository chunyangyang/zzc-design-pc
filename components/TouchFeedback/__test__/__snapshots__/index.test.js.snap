// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button 1`] = `
<div>
  <div
    class="touch-btn"
  >
    点击添加ClassName
  </div>
  <div
    class="touch-btn"
  >
    点击添加Style
  </div>
  <div
    class="touch-btn"
  >
    禁止触发touch
  </div>
</div>
`;
